### PR TITLE
Dockerizes all example projects

### PR DESCRIPTION
**Pull request by adriancole**
This takes from the docker setup @anuraaga made in sleuth-webmvc-example
and jigs it a little in preparation of multi-tag publications.

This punches a hole for supportsJoin as a couple APMs don't support that

See #45

### CODE REVIEW

File: docker/bin/start-brave-example [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    CLASS_NAME=Backend
+    ;;
+  * )
+    echo "Invalid argument: specify frontend or backend"
+    exit 1
+esac
+
+exec java ${JAVA_OPTS} -cp . \
+  -Dzipkin.endpoint=${ZIPKIN_ENDPOINT:=http://zipkin:9411/api/v2/spans} \
+  -Dzipkin.supportsJoin=${ZIPKIN_SUPPORTS_JOIN:=true} \
```
Comment (adriancole) [inline docker/bin/start-brave-example:17]: tested and this works

**Comment by adriancole** [comment]
If not deployed with our server, this image is just under 100MB
```
openzipkin/example-brave             webmvc4-boot          8eced5eb2942        13 minutes ago      99.7MB
```


If deployed with our server, it is an additional 12MB due to sharing a base layer
```
$ docker history openzipkin/example-brave:webmvc4-boot
IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
8eced5eb2942        13 minutes ago      /bin/sh -c #(nop) COPY dir:5bb676a79b3244855…   12MB                
56042738931c        13 minutes ago      /bin/sh -c #(nop)  ENTRYPOINT ["start-brave-…   0B                  
a50fc7dae6a5        13 minutes ago      /bin/sh -c #(nop)  EXPOSE 8081 9000             0B                  
d42307363664        13 minutes ago      /bin/sh -c #(nop) COPY dir:643ae2e6b07a50ab9…   532B                
feb44df65258        13 minutes ago      /bin/sh -c #(nop)  USER brave-example           0B                  
4c0a97c064bc        13 minutes ago      |1 USER=brave-example /bin/sh -c adduser -g …   4.66kB              
01b7d102ae02        13 minutes ago      /bin/sh -c #(nop) WORKDIR /brave-example        0B                  
04297f3aa5de        13 minutes ago      /bin/sh -c #(nop)  ARG USER=brave-example       0B                  
6eb44d6feb2f        13 minutes ago      /bin/sh -c #(nop)  ENV JAVA_OPTS=-Xms32m -Xm…   0B                  
73e9408f6226        15 hours ago        /bin/sh -c ln -s ${JAVA_HOME}/bin/java /usr/…   31B                 
<missing>           15 hours ago        /bin/sh -c #(nop) COPY dir:8535066c09bbd06ff…   81.5MB              
<missing>           15 hours ago        /bin/sh -c #(nop)  ENTRYPOINT ["/usr/bin/jav…   0B                  
<missing>           15 hours ago        /bin/sh -c #(nop)  ENV JAVA_HOME=/usr/lib/jv…   0B                  
<missing>           15 hours ago        /bin/sh -c apk add --no-cache libc6-compat      620kB               
<missing>           15 hours ago        /bin/sh -c echo 'hosts: files mdns4_minimal …   55B                 
<missing>           15 hours ago        /bin/sh -c #(nop)  ENV LANG=C.UTF-8             0B                  
<missing>           15 hours ago        /bin/sh -c #(nop)  MAINTAINER OpenZipkin "ht…   0B                  
<missing>           4 months ago        /bin/sh -c #(nop)  CMD ["/bin/sh"]              0B                  
<missing>           4 months ago        /bin/sh -c #(nop) ADD file:c92c248239f8c7b9b…   5.57MB              
```

**Comment by adriancole** [comment]
@anuraaga I've finished all the projects at this point. Just want to narrow down memory usage on the ones that aren't webmvc4-boot

Later, we can add a more modern example than Spring 4 or Spring Boot 1.5

File: docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+## Spring WebMVC 2.5/Servlet 2.5/Jetty 7/Java 6 example
+FROM azul/zulu-openjdk-alpine:6u119-6.22.0.3 as webmvc25
+
+WORKDIR /brave-example
+COPY --from=built /built/webmvc25/ .
+
+EXPOSE 8081 9000
+
+# When using JRE 1.6, we cannot use -XX:+ExitOnOutOfMemoryError
+ENV JAVA_OPTS="-Xms32m -Xmx64m -verbose:gc"
```
Comment (adriancole) [inline docker/Dockerfile:52]: verbose:gc here and below is temporary, while we check if non spring boot really needs this much memory

**Comment by adriancole** [comment]
images are done.. smaller ones are webmvc4 due to smaller JRE. `webmvc4-boot` is smaller, but requires more memory than `webmvc4`

```
openzipkin/example-brave   webmvc4-boot        a6c26e14cb24        About a minute ago   99.5MB
openzipkin/example-brave   webmvc4             bbfd260f05de        About a minute ago   116MB
openzipkin/example-brave   webmvc3             bc4b9354596b        About a minute ago   146MB
openzipkin/example-brave   webmvc25            cdadf8e03155        2 minutes ago        173MB
```

**Comment by adriancole** [comment]
glitch on webmvc4 suppose I need to add tests heh

```
$ docker logs frontend
Error: Could not find or load main class org.mortbay.jetty.runner.Runner
```

**Comment by adriancole** [comment]
ok all green. to tell if dockerhub automated builds will allow `DOCKER_BUILDKIT=1` or not will be easier to wait and see as opposed to anything else.

If Dockerhub automated builds fail on this, I'll add a travis config that adds the following the following setup when pushing on master:
```
{
  "experimental": false,
  "features": {
    "buildkit": true
  }
}
```

Someone who cares to change that later would be welcome to, just I'm a bit over time on this task and want to use known working setup (as I've had good luck on travis publication with the zipkin-java image)

File: .github/workflows/docker-tests.yml [hunk 3]
```diff
@@ -0,0 +1,44 @@
+name: Continuous Build Docker
+on:
+  push:
+    branches:
+      - main
```
Comment (anuraaga) [inline .github/workflows/docker-tests.yml:5]:
```suggestion
      - master
```

File: .github/workflows/docker-tests.yml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  pull_request:
```
Comment (anuraaga) [inline .github/workflows/docker-tests.yml:6]:
```suggestion
  pull_request:
    branches:
      - master
```

While we rarely have pull requests to other branches may as well stick to the branch

File: docker/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+WORKDIR /brave-example
+COPY --from=install /install/webmvc25-jetty/ .
+# Add HEALTHCHECK and ENTRYPOINT scripts into the default search path
+RUN mv docker-bin/* /usr/local/bin/ && rmdir docker-bin
+
+EXPOSE 8081 9000
+
+# When using JRE 1.6, we cannot use -XX:+ExitOnOutOfMemoryError
+ENV JAVA_OPTS="-Xms16m -Xmx16m"
```
Comment (anuraaga) [inline docker/Dockerfile:48]: Small app :)

File: docker/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+ENV JAVA_OPTS="-Xms16m -Xmx16m -XX:+ExitOnOutOfMemoryError"
+
+ENTRYPOINT ["start-brave-example"]
+
+# We use start period of 30s to avoid marking the container unhealthy on slow or contended CI hosts
+HEALTHCHECK --interval=1s --start-period=30s --timeout=5s CMD ["docker-healthcheck"]
+
+## Spring WebMVC 4/Servlet 3.1/Jetty 9/Java 8 example
+FROM openzipkin/java:1.8.0_252-b09-jre as webmvc4-jetty
```
Comment (anuraaga) [inline docker/Dockerfile:72]: Why Java 8?

**Review by anuraaga — APPROVED** [review]

File: docker/Dockerfile (hunk 6 above)
Comment (adriancole) [reply docker/Dockerfile:72]: I was intentionally trying to use the versions of java `brave.Platform` detects and were of the era of the example's dependencies. An upcoming more modern example could use latest!

**Comment by adriancole** [comment]
Thanks for the help! will 🤞 on master

**Pull request merged**